dot/state/grandpa.go

```diff
@@ -24,44 +26,570 @@ var (
 	currentSetIDKey   = []byte("setID")
 )
 
+var (
+	errPendingScheduledChanges = errors.New("pending scheduled changes needs to be applied")
+	errDuplicateHashes         = errors.New("duplicated hashes")
+	errAlreadyHasForcedChanges = errors.New("already has a forced change")
+	errUnfinalizedAncestor     = errors.New("ancestor with changes not applied")
+	errLowerThanBestFinalized  = errors.New("current finalized is lower than best finalized header")
+
+	ErrNoChanges = errors.New("cannot get the next authority change block number")
+)
+
+type pendingChange struct {
+	bestFinalizedNumber uint32
+	delay               uint32
+	nextAuthorities     []types.Authority
+	announcingHeader    *types.Header
+}
+
+func (p pendingChange) String() string {
+	return fmt.Sprintf("announcing header: %s (%d), delay: %d, next authorities: %d",
+		p.announcingHeader.Hash(), p.announcingHeader.Number, p.delay, len(p.nextAuthorities))
+}
+
+func (p *pendingChange) effectiveNumber() uint {
+	return p.announcingHeader.Number + uint(p.delay)
+}
+
+// appliedBefore compares the effective number between two pending changes
+// and returns true if:
+// - the current pending change is applied before the target
+// - the target is nil and the current contains a value
+func (p *pendingChange) appliedBefore(target *pendingChange) bool {
+	if p != nil && target != nil {
+		return p.effectiveNumber() < target.effectiveNumber()
+	}
+
+	return p != nil
+}
+
+type isDescendantOfFunc func(parent, child common.Hash) (bool, error)
+
+type pendingChangeNode struct {
+	change *pendingChange
+	nodes  []*pendingChangeNode
+}
+
+func (c *pendingChangeNode) importScheduledChange(blockHash common.Hash, blockNumber uint, pendingChange *pendingChange,
+	isDescendantOf isDescendantOfFunc) (imported bool, err error) {
+	announcingHash := c.change.announcingHeader.Hash()
+
+	if blockHash.Equal(announcingHash) {
+		return false, errDuplicateHashes
+	}
+
+	if blockNumber <= c.change.announcingHeader.Number {
+		return false, nil
+	}
```

Comment:
should this be an error? also, why do we need blockHash and blockNumber parameters if the pendingChange contains them?

Reply:
Not really, this is a recursive function so when the block number is less than or equal to the one which is persisted (`c.change`) then we should add the pending change as a sibling and not as a child of the `c.change`. I'm using separated params to hash an number values to become simple to read and avoid terms like: `pendingChange.announcingHeader.Number` or `pendingChange.announcingHeader.Hash()`

Reply:
ahh i see, i didn't see that it was recursive. that makes sense!